test/e2e_yaml_test.go

```diff
@@ -0,0 +1,238 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package test
+
+import (
+	"io"
+	"io/ioutil"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/clientcmd/api"
+	knativetest "knative.dev/pkg/test"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"testing"
+)
+
+const (
+	DEFAULT_KO_DOCKER_REPO = `gcr.io\/christiewilson-catfactory`
+	DEFAULT_NAMESPACE      = `namespace: default`
+)
+
+// Generalized for tekton CRDs
+func GetCreatedTektonCrd(input []byte, kind string) (string, error) {
+	re := regexp.MustCompile(kind + `.tekton.dev\/(.+) created`)
+	submatch := re.FindSubmatch(input)
+	if submatch == nil {
+		return "", nil
+	}
+	return string(submatch[1]), nil
+}
+
+func ValidatePipelineRunDone(t *testing.T, c *clients, pipelineRunName string) {
+	err := WaitForPipelineRunState(c, pipelineRunName, pipelineRunTimeout, Succeed(pipelineRunName), pipelineRunName)
+
+	if err != nil {
+		t.Fatal(err)
+	}
+	return
+}
+
+func ValidateTaskRunDone(t *testing.T, c *clients, taskRunName string) {
+	// Per test basis
+	err := WaitForTaskRunState(c, taskRunName, Succeed(taskRunName), taskRunName)
+
+	if err != nil {
+		t.Fatal(err)
+	}
+	return
+}
+
+// Substitute docker repos/bucket paths from environment to allow tests on
+// local clusters. Also update namespaces so that examples with ServiceAccounts
+// work under test
+func SubEnv(input []byte, namespace string) ([]byte, error) {
+	val, ok := os.LookupEnv("KO_DOCKER_REPO")
+	var output []byte
+	if ok {
+		re := regexp.MustCompile(DEFAULT_KO_DOCKER_REPO)
+		output = re.ReplaceAll(input, []byte(val))
+	} else {
+		return input, nil
+	}
+
+	re := regexp.MustCompile(DEFAULT_NAMESPACE)
+	output = re.ReplaceAll(output, []byte(strings.ReplaceAll(DEFAULT_NAMESPACE, "default", namespace)))
+	return output, nil
+}
+
+func KoCreate(input []byte, namespace string) ([]byte, error) {
+	cmd := exec.Command("ko", "create", "-n", namespace, "-f", "-")
+	stdin, err := cmd.StdinPipe()
+	if err != nil {
+		return []byte("Cannot create Stdin Pipe"), err
+	}
+
+	go func() {
+		defer stdin.Close()
+		io.WriteString(stdin, string(input))
+	}()
+
+	out, err := cmd.CombinedOutput()
+	return out, err
+}
+
+func DeleteClusterTask(t *testing.T, c *clients, name string) {
+	t.Logf("Deleting clustertask %s", name)
+	err := c.ClusterTaskClient.Delete(name, &metav1.DeleteOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func YamlSubtest(path string, validateFunc func(t *testing.T, c *clients, name string), kind string) func(subt *testing.T) {
+	return func(subt *testing.T) {
```

Comment:
this is a nitpick but i dont see much value in using `subt` vs `t` in this function - when i see `subt`in the code further down i wonder what it is, but if it was `t` then i would know exactly what it is - i know it would hide the `t` variable to use `t` for this - but maybe that's a good thing? (you probably dont want to interact with both?)